src/compiler/checker.ts

```diff
@@ -29514,14 +29512,53 @@ namespace ts {
+                    const basePropertyFlags = base.flags & SymbolFlags.PropertyOrAccessor;
+                    const derivedPropertyFlags = derived.flags & SymbolFlags.PropertyOrAccessor;
+                    if (basePropertyFlags && derivedPropertyFlags) {
+                        // property/accessor is overridden with property/accessor
+                        if (compilerOptions.legacyClassFields
+                            || baseDeclarationFlags & ModifierFlags.Abstract
+                            || base.valueDeclaration && base.valueDeclaration.parent.kind === SyntaxKind.InterfaceDeclaration
+                            || derived.valueDeclaration && isBinaryExpression(derived.valueDeclaration)) {
+                            // when the base property is abstract or from an interface, base/derived flags don't need to match
+                            // same when the derived property is from an assignment
+                            continue;
+                        }
+                        if (basePropertyFlags !== SymbolFlags.Property && derivedPropertyFlags === SymbolFlags.Property) {
+                            errorMessage = Diagnostics.Class_0_defines_instance_member_accessor_1_but_extended_class_2_defines_it_as_instance_member_property;
+                        }
+                        else if (basePropertyFlags === SymbolFlags.Property && derivedPropertyFlags !== SymbolFlags.Property) {
+                            errorMessage = Diagnostics.Class_0_defines_instance_member_property_1_but_extended_class_2_defines_it_as_instance_member_accessor;
+                        }
+                        else {
+                            const uninitialized = find(derived.declarations, d => d.kind === SyntaxKind.PropertyDeclaration && !(d as PropertyDeclaration).initializer);
+                            if (uninitialized
+                                && !(derived.flags & SymbolFlags.Transient)
+                                && !(baseDeclarationFlags & ModifierFlags.Abstract)
+                                && !(derivedDeclarationFlags & ModifierFlags.Abstract)
+                                && !derived.declarations.some(d => d.flags & NodeFlags.Ambient)) {
+                                const constructor = findConstructorDeclaration(getClassLikeDeclarationOfSymbol(type.symbol)!);
+                                const propName = (uninitialized as PropertyDeclaration).name;
+                                if ((uninitialized as PropertyDeclaration).exclamationToken
+                                    || !constructor
+                                    || !isIdentifier(propName)
+                                    || !strictNullChecks
+                                    || !isPropertyInitializedInConstructor(propName, type, constructor)) {
+                                    const errorMessage = Diagnostics.Property_0_will_overwrite_the_base_property_in_1_Add_a_declare_modifier_or_an_initializer_to_avoid_this;

```

Comment:
> Add_a_declare_modifier_or_an_initializer_to_avoid_this

Suggestion: if the base property’s type is identical to the derived property’s type, better advice might be to delete the property declaration altogether.

Reply:
I don't know; I saw enough of this that I think there must be *some* kind of intent behind it. I don't think people truly trust inheritance. I know I don't.

Reply:
Although adding an initializer isn't a propos in this case.

Reply:
Right—adding an an initializer changes the semantics, so assuming you don’t want that, the guidance is to add `declare`, which is utterly meaningless, no?

Reply:
I’m not saying that “delete this” must be the _only_ option listed in the error message; maybe

> Property {0} will overwrite the base property in {1}. To silence this message, add an initializer. To avoid overwriting the base property, add a 'declare' modifier or remove this redundant declaration.

Reply:
How about:
> "Property '{0}' will overwrite the base property in '{1}'. If this is intentional, add an initializer. Otherwise, add a 'declare' modifier or remove the redundant declaration.": {

Reply:
Sounds good. I like that it clarifies that adding an initializer and adding a declare modifier essentially have _opposite_ semantics. The original error message kind of sounds like they’ll have the same effect.